File: bender/bender-kt/src/main/java/com/github/dnaka91/bender/kt/Mandelbrot.kt
package com.github.dnaka91.bender.kt

import org.apache.commons.math3.complex.Complex
import kotlin.math.pow

internal object Mandelbrot {
    fun render(width: Int, height: Int): ByteArray {
        val pixels = ByteArray(width * height)
        val upperLeft = Complex(-1.2, 0.35)
        val lowerRight = Complex(-1.0, 0.2)

        for (row in 0 until height) {
            for (column in 0 until width) {
                val point = pxToPoint(width, height, column, row, upperLeft, lowerRight)

                pixels[row * width + column] = escape(point, 255).let { 255 - it }.toByte()
            }
        }

        return pixels
    }

    private fun escape(c: Complex, limit: Int): Int {
        var z = Complex(0.0, 0.0)
        for (i in 0..limit) {
            z = z.multiply(z).add(c)
            if (z.real.pow(2) + z.imaginary.pow(2) > 4.0) {
                return i
            }
        }

        return 255
    }

    private fun pxToPoint(
        bounds_w: Int,
        bounds_h: Int,
        x: Int,
        y: Int,
        upperLeft: Complex,
        lowerRight: Complex
    ): Complex {
        val width = lowerRight.real - upperLeft.real
        val height = upperLeft.imaginary - lowerRight.imaginary

        return Complex(
            upperLeft.real + x.toDouble() * width / bounds_w.toDouble(),
            upperLeft.imaginary - y.toDouble() * height / bounds_h.toDouble()
        )
    }
}

File: bender/bend/uniffi/src/mandelbrot.rs
use bend::{
    mandelbrot::{self, Complex64},
    mandelbrot_manual,
};

pub fn mandelbrot(width: u32, height: u32) -> Vec<u8> {
    mandelbrot::render(
        (width, height),
        Complex64 {
            re: -1.20,
            im: 0.35,
        },
        Complex64 {
            re: -1.00,
            im: 0.20,
        },
    )
}

pub fn mandelbrot_manual(width: u32, height: u32) -> Vec<u8> {
    mandelbrot_manual::render((width, height))
}
